Windows SDK Generates C4668 During Pre-processing - Confuses Sccache

## Issue

Sccache cannot compile some Windows applications via MSVC when `/we4668` is specified because first it tries to use the same arguments to preprocess the files, which fails within the Windows SDK.

Specifically, when trying to use sccache to cache Unreal Engine compilation (after fixing some other issues) sccache produces the error `preprocessor failed` from: https://github.com/mozilla/sccache/blob/main/src/compiler/c.rs#L339

In the include list in this error, the relevant error is visible:
```
Note: including file: C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared\intsafe.h
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared\intsafe.h(66): error C4668: '__midl' is not defined as a preprocessor macro, replacing with '0' for '#if/#elif'
```

## Why is this a problem with sccache?

The reason sccache runs into this problem, but Unreal Engine does not, is that sccache faithfully copies all arguments for compilation into the preprocessor command it executes. However, when Unreal Engine builds, it explicitly disables this error message when preprocessing and only enables it during compiling.

## Expected behaviour

I believe the correct behaviour here is that this compiler warning should be explicitly ignored when building using MSVC.

Sample output from the above:

```pwsh
PS ~\git\reproduction-of-sccache-windows-sdk-issue> $compile = "-c", "main.cpp", "-nologo", "/IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\INCLUDE", "/IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8.1\include\um", "/IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt", "/IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared", "/we4668", "/D__STDC_WANT_SECURE_LIB__=1", "/EHsc"
PS ~\git\reproduction-of-sccache-windows-sdk-issue> cl.exe @compile
main.cpp
PS ~\git\reproduction-of-sccache-windows-sdk-issue> ..\sccache\target\debug\sccache.exe cl.exe @compile
main.cpp
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared\intsafe.h(66): error C4668: '__midl' is not defined as a preprocessor macro, replacing with '0' for '#if/#elif'
```

--- a/.github/workflows/integration-tests.yml
+++ b/.github/workflows/integration-tests.yml
@@ -400,6 +400,13 @@ jobs:
           test -e ./foo.o || { echo "No compiler output found"; exit -1; }
           test -e ./foo.o.json || { echo "No dependency list found"; exit -1; }
 
+      - name: Compile - Preprocessing Compiler Bug
+        shell: bash
+        working-directory: ./tests/msvc-preprocessing
+        run: |
+          $SCCACHE_EXE "$(where cl.exe)" -c "@args.rsp"
+          $SCCACHE_EXE --show-stats
+
       - name: Stop Server
         if: success() || failure()
         shell: bash

--- a/src/compiler/msvc.rs
+++ b/src/compiler/msvc.rs
@@ -950,9 +950,8 @@ where
         .args(&parsed_args.dependency_args)
         .args(&parsed_args.common_args)
         // Windows SDK generates C4668 during preprocessing, but compiles fine.
-        // Making sure that 4668 is a level 4 warning prevents failing preprocessing on it.
         // Read for more info: https://github.com/mozilla/sccache/issues/1725
-        .arg("/w44668")
+        .arg("/wd4668")
         .env_clear()
         .envs(env_vars.iter().map(|&(ref k, ref v)| (k, v)))
         .current_dir(cwd);

--- a/src/compiler/msvc.rs
+++ b/src/compiler/msvc.rs
@@ -949,6 +949,10 @@ where
         .args(&parsed_args.preprocessor_args)
         .args(&parsed_args.dependency_args)
         .args(&parsed_args.common_args)
+        // Windows SDK generates C4668 during preprocessing, but compiles fine.
+        // Making sure that 4668 is a level 4 warning prevents failing preprocessing on it.
+        // Read for more info: https://github.com/mozilla/sccache/issues/1725
+        .arg("/w44668")
         .env_clear()
         .envs(env_vars.iter().map(|&(ref k, ref v)| (k, v)))
         .current_dir(cwd);

--- a/tests/msvc-preprocessing/args.rsp
+++ b/tests/msvc-preprocessing/args.rsp
@@ -1,0 +1,1 @@
+foo.cpp -Fofoo.o -W4 -WX -Wall

--- a/tests/msvc-preprocessing/foo.cpp
+++ b/tests/msvc-preprocessing/foo.cpp
@@ -1,0 +1,15 @@
+// This tests sccache's ability to handle a known bug in cl.exe
+// More information: https://github.com/mozilla/sccache/issues/1725
+__pragma(warning(push))
+__pragma(warning(disable: 4668))
+// cl.exe during compilation will correctly ignore 4668 here (undefined define)
+// during preprocessing, it will not ignore 4668
+// sccache must explicitly ignore this warning when preprocessing (to figure out the cache key)
+#if UNDEFINED_MACRO_TRIGGERING_C4668
+#error "This error should be unreachable"
+#endif
+__pragma(warning(pop))
+
+int main() {
+  return 0;
+}

